xmtp_mls/src/groups/device_sync/backup/backup_exporter.rs

```diff
@@ -0,0 +1,158 @@
+            nonce_buffer,
+        }
+    }
+}
+
+impl AsyncRead for BackupExporter {
+    /// This function encrypts first, and compresses second.
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.get_mut();
+
+        // Putting this up here becuase we don't want to encrypt or compress the nonce.
+        if matches!(this.stage, Stage::Nonce) {
+            let amount = this.nonce_buffer.len().min(buf.len());
+            let nonce_bytes: Vec<_> = this.nonce_buffer.drain(..amount).collect();
+            buf[..amount].copy_from_slice(&nonce_bytes);
+
+            if this.nonce_buffer.is_empty() {
+                this.stage = Stage::Metadata;
+            }
+            return Poll::Ready(Ok(amount));
+        }
+
+        {
+            // Read from the buffer while there is data
+            let buffer_inner = this.zstd_encoder.get_ref();
+            if this.position < buffer_inner.len() {
+                let available = &buffer_inner[this.position..];
+                let amount = available.len().min(buf.len());
+                buf[..amount].copy_from_slice(&available[..amount]);
+                this.position += amount;
+
+                return Poll::Ready(Ok(amount));
+            }
+        }
+
+        // The buffer is consumed. Reset.
+        this.position = 0;
+        this.zstd_encoder.get_mut().clear();
+
+        // Time to fill the buffer with more data 8kb at a time.
+        while this.zstd_encoder.get_ref().len() < 8_000 {
+            let element = match this.stage {
+                Stage::Nonce => {
+                    // Should never get here due to the above logic. Error if it does.
+                    unreachable!()
+                }
+                Stage::Metadata => {
+                    this.stage = Stage::Elements;
+                    BackupElement {
+                        element: Some(Element::Metadata(this.metadata.clone())),
+                    }
+                    .encode_to_vec()
+                }
+                Stage::Elements => match this.stream.poll_next_unpin(cx) {
+                    Poll::Ready(Some(element)) => element.encode_to_vec(),
+                    Poll::Pending => return Poll::Pending,
+                    Poll::Ready(None) => {
+                        if !this.encoder_finished {
+                            this.encoder_finished = true;
+                            let fut = this.zstd_encoder.close();
+                            pin_mut!(fut);
+                            let _ = fut.poll(cx)?;
+                        }
+                        break;
+                    }
+                },
+            };
+
+            let mut element = this
+                .cipher
+                .encrypt(&this.nonce, &*element)
+                .expect("Encryption should always work");
+            let mut bytes = (element.len() as u32).to_le_bytes().to_vec();
```

Comment:
what's the reasoning for little endian here?

Reply:
Compatibility across architectures, even though x86 and arm are both little endian, it's nice to be safe.